vm/devices/storage/storvsc_driver/src/lib.rs

```diff
@@ -0,0 +1,913 @@
+            Ok(_) => {}
+            Err(err) => tracing::error!(error = err.as_error(), "storvsc run error"),
+        }
+        Ok(())
+    }
+}
+
+impl<T: 'static + Send + Sync + RingMem> InspectTask<Storvsc<T>> for StorvscState {
+    fn inspect(&self, req: inspect::Request<'_>, worker: Option<&Storvsc<T>>) {
+        if let Some(worker) = worker {
+            let mut resp = req.respond();
+            resp.field("has_negotiated", worker.has_negotiated);
+        }
+    }
+}
+
+impl<T: 'static + Send + Sync + RingMem> Storvsc<T> {
+    pub(crate) fn new(
+        channel: RawAsyncChannel<T>,
+        version: storvsp_protocol::ProtocolVersion,
+        new_request_receiver: Receiver<StorvscRequest>,
+    ) -> Result<Self, StorvscError> {
+        let queue =
+            Queue::new(channel).map_err(|err| StorvscError(StorvscErrorInner::Queue(err)))?;
+
+        Ok(Self {
+            inner: StorvscInner {
+                new_request_receiver,
+                transactions: Slab::new(),
+            },
+            version,
+            queue,
+            num_sub_channels: None,
+            has_negotiated: false,
+        })
+    }
+}
+
+impl<T: Send + Sync + RingMem> Storvsc<T> {
+    async fn negotiate(&mut self) -> Result<(), StorvscError> {
+        // Negotiate protocol with storvsp instance on the other end of VMBus
+        // Step 1: BEGIN_INITIALIZATION
+        self.inner
+            .send_packet_and_expect_completion(
+                &mut self.queue,
+                storvsp_protocol::Operation::BEGIN_INITIALIZATION,
+                1,
+                &(),
+            )
+            .await?;
+
+        // Step 2: QUERY_PROTOCOL_VERSION - request latest version
+        self.inner
+            .send_packet_and_expect_completion(
+                &mut self.queue,
+                storvsp_protocol::Operation::QUERY_PROTOCOL_VERSION,
+                2,
+                &self.version,
+            )
+            .await
+            .map_err(|err| match err {
+                StorvscError(StorvscErrorInner::PacketError(PacketError::UnexpectedStatus(
+                    storvsp_protocol::NtStatus::INVALID_DEVICE_STATE,
+                ))) => StorvscError(StorvscErrorInner::UnsupportedProtocolVersion),
+                _ => err,
+            })?;
+
+        // Step 3: QUERY_PROPERTIES
+        let properties_packet = self
+            .inner
+            .send_packet_and_expect_completion(
+                &mut self.queue,
+                storvsp_protocol::Operation::QUERY_PROPERTIES,
+                3,
+                &(),
+            )
+            .await?;
+        let _properties = storvsp_protocol::ChannelProperties::ref_from_prefix(
+            &properties_packet.data[0..properties_packet.data_size],
+        )
+        .map_err(|_err| StorvscError(StorvscErrorInner::UnexpectedOperation))?
+        .0
+        .to_owned();
+
+        // Skip subchannels because unsupported at the moment
```

Comment:
Presumably we'll want to eventually add support for subchannels here. Please file an item on the backlog for that.